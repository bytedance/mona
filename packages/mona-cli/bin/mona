#! /usr/bin/env node

const mona = require('../dist').default;
mona();
