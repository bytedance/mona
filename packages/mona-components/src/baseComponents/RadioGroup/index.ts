import React from 'react';
import createComponent from '../../createComponent';
import { BaseProps, EventHandler } from '../base';

export interface RadioGroupProps extends BaseProps {
  onChange?: EventHandler;
  name?: string;
}

export const RadioGroup: React.ComponentType<RadioGroupProps> = createComponent<RadioGroupProps>('radio-group')

