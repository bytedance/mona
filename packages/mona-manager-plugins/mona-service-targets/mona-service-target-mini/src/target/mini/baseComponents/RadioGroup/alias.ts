import { bubbleEventsAlias, basePropAlias, propAliasMap } from '../prop';
import { RadioGroupAlias as Alias } from '../type';

const alias: Alias = {
  ...basePropAlias,
  ...bubbleEventsAlias,
  name: propAliasMap.name,
  onChange: propAliasMap.onChange,
};

export default alias;
