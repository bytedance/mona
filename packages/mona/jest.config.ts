import baseConfig from '../../jestconfig.base';

module.exports = baseConfig;
