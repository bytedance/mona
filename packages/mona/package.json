{
  "name": "@bytedance/mona",
  "version": "0.2.0-alpha.9",
  "description": "Merchant app's developing and building tools",
  "main": "./dist/index",
  "scripts": {
    "build": "yarn clean && ttsc -b",
    "start": "ttsc -b -w",
    "clean": "rm -rf dist",
    "test": "jest"
  },
  "files": [
    "dist"
  ],
  "author": "zhangxiaowei.chris@bytedance.com",
  "license": "MIT",
  "publishConfig": {
    "access": "public",
    "registry": "https://registry.npmjs.org"
  },
  "devDependencies": {
    "copy-webpack-plugin": "^10.2.0",
    "react": "^17.0.2",
    "webpack": "^5.58.2",
    "webpack-chain": "^6.5.1"
  },
  "peerDependencies": {
    "react": "^17.0.2"
  },
  "gitHead": "cc5aa5283af595fd7d8a7b4921f54e2797504edd"
}
