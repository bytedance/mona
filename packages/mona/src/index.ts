export { createProjectConfig, ProjectConfig } from './createProjectConfig';
export { createAppConfig, AppConfig } from './createAppConfig';
export { createPageConfig, PageConfig } from './createPageConfig';
export { default as BaseApis } from './BaseApis';
export * from './BaseApis';
export { default as BaseComponents } from './BaseComponents';
export * from './BaseComponents';
export * from './types';
export * from './lifecycle';
