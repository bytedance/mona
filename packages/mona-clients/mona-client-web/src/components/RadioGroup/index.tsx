import React, { useCallback, useRef } from 'react';
import { RadioGroupProps, TouchEvent } from '@bytedance/mona';
import { useHandlers } from '../hooks';
import { useFormContext } from '../Form/hooks';

interface RadioGroupContextProps {
  initValue: (value?: string, checked?: boolean, clear?: () => void) => number;
  changeValue: (index: number, value?: string) => void
  toggleChecked: (index: number, checked?: boolean, e?: TouchEvent) => void
}

interface ValueItem {
  value?: string;
  checked?: boolean;
}

export const RadioGroupContext = React.createContext<RadioGroupContextProps | null>(null)

const RadioGroup: React.FC<RadioGroupProps> = ({ children, name = '', onChange, ...restProps }) => {
  const { handleClassName, ...handlerProps } = useHandlers(restProps);
  const valuesRef = useRef<ValueItem[]>([])
  const indexRef = useRef(0);
  const clearsRef = useRef<((() => void) | undefined)[]>([]);

  const reset = useCallback(() => {
    const newValues = valuesRef.current;
    const cbs = clearsRef.current;
    for (let i = 0; i < newValues.length; i++) {
      newValues[i] = { ...newValues[i], checked: false }
      const cb = cbs[i];
      if (typeof cb === 'function') {
        cb();
      }
    }
  }, [clearsRef, valuesRef]);
  const getValues = () => {
    return valuesRef.current.filter(v => v.checked).map(v => v.value)[0] || '';
  }
  useFormContext(name, getValues, reset);
  
  const context = {
    initValue: (value?: string, checked?: boolean, clear?: () => void) => {
      const index = indexRef.current++;
      const newValues = [...valuesRef.current];
      // record order
      newValues[index] = { value, checked };
      valuesRef.current = newValues;
      clearsRef.current[index] = clear;
      return index;
    },
    changeValue: (index: number, value?: string) => {
      const newValues = [...valuesRef.current];
      newValues[index] = { ...newValues[index], value };
      valuesRef.current = newValues;
    },
    toggleChecked: (index: number, checked?: boolean, event?: TouchEvent) => {
      const newValues = [...valuesRef.current];
      const cbs = clearsRef.current;

      for (let i = 0; i < newValues.length; i++) {
        if (index !== i) {
          newValues[i] = { ...newValues[i], checked: false }
          const cb = cbs[i];
          if (checked && typeof cb === 'function') {
            cb();
          }
        } else {
          newValues[i] = { ...newValues[i], checked }
        }
      }

      valuesRef.current = newValues;
      
      if (typeof onChange === 'function') {
        const value = newValues[index];
        const e = { ...event!, detail: { value: value.checked ? value.value : undefined } };
        onChange(e)
      }
    }
  }

  return (
    <div className={handleClassName()} {...handlerProps}>
      <RadioGroupContext.Provider value={context}>
        {children}
      </RadioGroupContext.Provider>
    </div>
  )
}

export default RadioGroup;
