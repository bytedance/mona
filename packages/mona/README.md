# @bytedance/mona
对外暴露提供构建时API
## 配置创建
### createProjectConfig
创建项目配置，`mona.config.ts`中使用
#### 例子
```js
// mona.config.ts
import { createProjectConfig } from '@bytedance/mona';
export default createProjectConfig({
  projectName: 'demo',
  input: './src/app.tsx',
  output: 'dist'
})
```

#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| projectName  | 应用/插件名称 | 是 | string | - |
| input  | 入口文件路径 | 否 | string | - |
| output  | 打包后的文件目录路径 | 否 | string | - |
| raw  | 定义函数来修改webpack配置 | 否 | (options: any) => any | - |
| dev  | 用来修改本地服务配置 | 否 | string | { port: 9999 } |

### createAppConfig
创建应用配置，`app.config.ts`中使用
#### 例子
```js
// app.config.ts
import { createAppConfig } from '@bytedance/mona';
export default createAppConfig({
  pages: [
    'pages/Home/index',
    'pages/Info/index',
    'pages/List/index'
  ]
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| pages  | 页面路径 | 是 | string[] | - |

默认`pages`的第一项为首页
### createPageConfig
创建页面配置，需在每个页面目录的`page.config.ts`中使用

#### 例子
```js
// page.config.ts
import { createPageConfig } from '@bytedance/mona';
export default createPageConfig({
  navigationBarTitleText: 'page title'
})
```
#### 参数说明
|  参数   | 说明  | 是否必填 | 类型 | 默认值 |
|  ----  | ----  | ---- | ---- | ---- |
| navigationBarTitleText  | 当前页面标题 | 否 | string | - |
