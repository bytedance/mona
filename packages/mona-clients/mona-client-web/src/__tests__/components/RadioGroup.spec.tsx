import React from 'react';
import { mount } from 'enzyme';
import RadioGroup from '../../components/RadioGroup';
import mountTest from '../../../../../tests/shared/mountTest';
import handlerTest from '../../../../../tests/shared/handlerTest';
import { RadioGroupProps } from '@bytedance/mona';

describe('web component: RadioGroup', () => {
  mountTest(RadioGroup)
  handlerTest({ Component: RadioGroup, firstTag: 'div' });

  it('should render correctly', () => {
    expect(mount(<RadioGroup />).render()).toMatchSnapshot();
  })
  
  it('should have right properties', () => {
    const props: RadioGroupProps = {
      name: 'hello',
      className: 'class-name',
      style: { color: 'red' },
    }
    const wrapper = mount(<RadioGroup {...props}>hello</RadioGroup>);
    expect(wrapper.render()).toMatchSnapshot();
  })
})
