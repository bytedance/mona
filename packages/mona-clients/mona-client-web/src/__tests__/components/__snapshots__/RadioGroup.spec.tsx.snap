// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web component: RadioGroup should have right properties 1`] = `
LoadedCheerio {
  "0": <div
    class="class-name"
    style="color: red;"
  >
    hello
  </div>,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;

exports[`web component: RadioGroup should render correctly 1`] = `
LoadedCheerio {
  "0": <div />,
  "_root": LoadedCheerio {
    "0": Document {
      "children": Array [
        <html>
          <head />
          <body />
        </html>,
      ],
      "endIndex": null,
      "next": null,
      "parent": null,
      "prev": null,
      "startIndex": null,
      "type": "root",
      "x-mode": "quirks",
    },
    "_root": [Circular],
    "length": 1,
    "options": Object {
      "decodeEntities": true,
      "xml": false,
    },
  },
  "length": 1,
  "options": Object {
    "decodeEntities": true,
    "xml": false,
  },
}
`;
